benches/benchmarks.rs

```diff
@@ -0,0 +1,292 @@
+    group.throughput(Throughput::Bytes(1));
+
+    group.bench_function("ascii_chars", |b| {
+        b.iter(|| BIG_16.find(&haystack));
+    });
+    group.bench_function("stdlib_find_char_set", |b| {
+        b.iter(|| {
+            haystack.find(
+                &[
+                    'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P',
+                ][..],
+            )
+        });
+    });
+    group.bench_function("stdlib_find_closure", |b| {
+        b.iter(|| {
+            haystack.find(|c| {
+                c == 'A'
+                    || c == 'B'
+                    || c == 'C'
+                    || c == 'D'
+                    || c == 'E'
+                    || c == 'F'
+                    || c == 'G'
+                    || c == 'H'
+                    || c == 'I'
+                    || c == 'J'
+                    || c == 'K'
+                    || c == 'L'
+                    || c == 'M'
+                    || c == 'N'
+                    || c == 'O'
+                    || c == 'P'
+            })
+        });
+    });
+    group.bench_function("stdlib_iter_position", |b| {
+        b.iter(|| {
+            haystack.bytes().position(|c| {
+                c == b'A'
+                    || c == b'B'
+                    || c == b'C'
+                    || c == b'D'
+                    || c == b'E'
+                    || c == b'F'
+                    || c == b'G'
+                    || c == b'H'
+                    || c == b'I'
+                    || c == b'J'
+                    || c == b'K'
+                    || c == b'L'
+                    || c == b'M'
+                    || c == b'N'
+                    || c == b'O'
+                    || c == b'P'
+            })
+        });
+    });
+    group.bench_function("memchr", |b| {
+        b.iter(|| {
+            let bytes = haystack.as_bytes();
+            let indexes = [
+                memchr::memchr3(b'A', b'B', b'C', bytes),
+                memchr::memchr3(b'D', b'E', b'F', bytes),
+                memchr::memchr3(b'G', b'H', b'I', bytes),
+                memchr::memchr3(b'J', b'K', b'L', bytes),
+                memchr::memchr3(b'M', b'N', b'O', bytes),
+                memchr::memchr(b'P', bytes),
+            ];
+            indexes.iter().copied().flatten().min()
+        })
+    });
+}
+
+fn substr(c: &mut Criterion) {
+    let mut haystack = prefix_string();
+    haystack.push_str("xyzzy");
+    let haystack = black_box(haystack);
+
+    let mut group = c.benchmark_group("find_substring");
+    group.throughput(Throughput::Bytes(haystack.len() as u64));
+
+    group.bench_function("substring", |b| {
+        b.iter(|| SUBSTRING.find(&haystack));
+    });
+    group.bench_function("stdlib_find_string", |b| {
+        b.iter(|| haystack.find("xyzzy"));
+    });
+    group.bench_function("memchr", |b| {
+        b.iter(|| memchr::memmem::find(haystack.as_bytes(), b"xyzzy"));
```

Comment:
Note that in the jetscii case, it looks like you're building a searcher once and then only benchmarking the search function. But in the memchr case, you're benchmarking both searcher construction and the search. You'll want to [build a `Finder`](https://docs.rs/memchr/latest/memchr/memmem/struct.Finder.html) once, like you build a `SUBSTRING` once.

(Both cases are interesting to measure. Note that `std` doesn't provide a way to build the searcher once and reuse it with different haystacks.)

Reply:
I did actually originally do that, but since memchr was so far ahead on the m1 mac when I fisrt wrote it (fallback vs new memchr simd) I thought it didn't need that help

Reply:
It's a less a question of "need the help" and more just about comparing apples-to-apples where possible. The haystack (5KB I think?) is small enough here that it might actually impact the timings in a meaningful way. Don't know for sure though.

Reply:
I made the change to re-use a Finder.

> 5KB I think?

It's a 5MiB haystack, which probably explains why Criteron said it was within the margin of error.

While I have you, is there any chance you could look at https://github.com/tafia/quick-xml/pull/664 if you have any ideas for making xml escaping faster?